File: app/src/main/java/com/atguigu/p2p/ui/randomLayout/AnimationUtil.java
package com.atguigu.p2p.ui.randomLayout;

import android.view.animation.AlphaAnimation;
import android.view.animation.Animation;
import android.view.animation.AnimationSet;
import android.view.animation.DecelerateInterpolator;
import android.view.animation.LinearInterpolator;
import android.view.animation.ScaleAnimation;

public class AnimationUtil {

    private static final long MEDIUM = 500;

    /**
     * 创建一个淡入放大的动画
     */
    public static Animation createZoomInNearAnim() {
        AnimationSet ret;
        Animation anim;
        ret = new AnimationSet(false);
        // 创建一个淡入的动画
        anim = new AlphaAnimation(0f, 1f);
        anim.setDuration(MEDIUM);
        anim.setInterpolator(new LinearInterpolator());
        ret.addAnimation(anim);
        // 创建一个放大的动画
        anim = new ScaleAnimation(0, 1, 0, 1, Animation.RELATIVE_TO_SELF, 0.5f, Animation.RELATIVE_TO_SELF, 0.5f);
        anim.setDuration(MEDIUM);
        anim.setInterpolator(new DecelerateInterpolator());
        ret.addAnimation(anim);
        return ret;
    }

    /**
     * 创建一个淡出放大的动画
     */
    public static Animation createZoomInAwayAnim() {
        AnimationSet ret;
        Animation anim;
        ret = new AnimationSet(false);
        // 创建一个淡出的动画
        anim = new AlphaAnimation(1f, 0f);
        anim.setDuration(MEDIUM);
        anim.setInterpolator(new DecelerateInterpolator());
        ret.addAnimation(anim);
        // 创建一个放大的动画
        anim = new ScaleAnimation(1, 3, 1, 3, Animation.RELATIVE_TO_SELF, 0.5f, Animation.RELATIVE_TO_SELF, 0.5f);
        anim.setDuration(MEDIUM);
        anim.setInterpolator(new DecelerateInterpolator());
        ret.addAnimation(anim);
        return ret;
    }

    /**
     * 创建一个淡入缩小的动画
     */
    public static Animation createZoomOutNearAnim() {
        AnimationSet ret;
        Animation anim;
        ret = new AnimationSet(false);
        // 创建一个淡入的动画
        anim = new AlphaAnimation(0f, 1f);
        anim.setDuration(MEDIUM);
        anim.setInterpolator(new LinearInterpolator());
        ret.addAnimation(anim);
        // 创建一个缩小的动画
        anim = new ScaleAnimation(3, 1, 3, 1, Animation.RELATIVE_TO_SELF, 0.5f, Animation.RELATIVE_TO_SELF, 0.5f);
        anim.setDuration(MEDIUM);
        anim.setInterpolator(new DecelerateInterpolator());
        ret.addAnimation(anim);
        return ret;
    }

    /**
     * 创建一个淡出缩小的动画
     */
    public static Animation createZoomOutAwayAnim() {
        AnimationSet ret;
        Animation anim;
        ret = new AnimationSet(false);
        // 创建一个淡出的动画
        anim = new AlphaAnimation(1f, 0f);
        anim.setDuration(MEDIUM);
        anim.setInterpolator(new DecelerateInterpolator());
        ret.addAnimation(anim);
        // 创建一个缩小的动画
        anim = new ScaleAnimation(1, 0, 1, 0, Animation.RELATIVE_TO_SELF, 0.5f, Animation.RELATIVE_TO_SELF, 0.5f);
        anim.setDuration(MEDIUM);
        anim.setInterpolator(new DecelerateInterpolator());
        ret.addAnimation(anim);
        return ret;
    }

    /**
     * 创建一个淡入放大的动画
     */
    public static Animation createPanInAnim(float degree) {
        AnimationSet ret;
        Animation anim;
        ret = new AnimationSet(false);
        // 创建一个淡入动画
        anim = new AlphaAnimation(0f, 1f);
        anim.setDuration(MEDIUM);
        anim.setInterpolator(new LinearInterpolator());
        ret.addAnimation(anim);
        // 创建一个放大动画
        final float pivotX = (float) (1 - Math.cos(degree)) / 2;
        final float pivotY = (float) (1 + Math.sin(degree)) / 2;

        anim = new ScaleAnimation(0.8f, 1, 0.8f, 1, Animation.RELATIVE_TO_SELF, pivotX, Animation.RELATIVE_TO_SELF,
                pivotY);
        anim.setDuration(MEDIUM);
        anim.setInterpolator(new DecelerateInterpolator());
        ret.addAnimation(anim);

        return ret;
    }

    /**
     * 创建一个淡出缩小的动画
     */
    public static Animation createPanOutAnim(float degree) {
        AnimationSet ret;
        Animation anim;
        ret = new AnimationSet(false);
        // 创建一个淡出动画
        anim = new AlphaAnimation(1f, 0f);
        anim.setDuration(MEDIUM);
        anim.setInterpolator(new DecelerateInterpolator());
        ret.addAnimation(anim);
        // 创建一个缩小动画
        final float pivotX = (float) (1 + Math.cos(degree)) / 2;
        final float pivotY = (float) (1 - Math.sin(degree)) / 2;
        anim = new ScaleAnimation(1, 0.8f, 1, 0.8f, Animation.RELATIVE_TO_SELF, pivotX, Animation.RELATIVE_TO_SELF,
                pivotY);
        anim.setDuration(MEDIUM);
        anim.setInterpolator(new DecelerateInterpolator());
        ret.addAnimation(anim);

        return ret;
    }
}

File: app/src/main/java/com/xpf/p2p/widget/randomLayout/AnimationUtil.kt
package com.xpf.p2p.widget.randomLayout

import android.view.animation.AlphaAnimation
import android.view.animation.Animation
import android.view.animation.AnimationSet
import android.view.animation.DecelerateInterpolator
import android.view.animation.LinearInterpolator
import android.view.animation.ScaleAnimation

object AnimationUtil {

    private const val MEDIUM = 500L

    @JvmStatic
    fun createZoomInNearAnim(): Animation {
        val ret = AnimationSet(false)
        var anim: Animation = AlphaAnimation(0f, 1f)
        anim.duration = MEDIUM
        anim.interpolator = LinearInterpolator()
        ret.addAnimation(anim)
        anim = ScaleAnimation(0f, 1f, 0f, 1f, Animation.RELATIVE_TO_SELF, 0.5f, Animation.RELATIVE_TO_SELF, 0.5f)
        anim.duration = MEDIUM
        anim.interpolator = DecelerateInterpolator()
        ret.addAnimation(anim)
        return ret
    }

    @JvmStatic
    fun createZoomInAwayAnim(): Animation {
        val ret = AnimationSet(false)
        var anim: Animation = AlphaAnimation(1f, 0f)
        anim.duration = MEDIUM
        anim.interpolator = DecelerateInterpolator()
        ret.addAnimation(anim)
        anim = ScaleAnimation(1f, 3f, 1f, 3f, Animation.RELATIVE_TO_SELF, 0.5f, Animation.RELATIVE_TO_SELF, 0.5f)
        anim.duration = MEDIUM
        anim.interpolator = DecelerateInterpolator()
        ret.addAnimation(anim)
        return ret
    }

    @JvmStatic
    fun createZoomOutNearAnim(): Animation {
        val ret = AnimationSet(false)
        var anim: Animation = AlphaAnimation(0f, 1f)
        anim.duration = MEDIUM
        anim.interpolator = LinearInterpolator()
        ret.addAnimation(anim)
        anim = ScaleAnimation(3f, 1f, 3f, 1f, Animation.RELATIVE_TO_SELF, 0.5f, Animation.RELATIVE_TO_SELF, 0.5f)
        anim.duration = MEDIUM
        anim.interpolator = DecelerateInterpolator()
        ret.addAnimation(anim)
        return ret
    }

    @JvmStatic
    fun createZoomOutAwayAnim(): Animation {
        val ret = AnimationSet(false)
        var anim: Animation = AlphaAnimation(1f, 0f)
        anim.duration = MEDIUM
        anim.interpolator = DecelerateInterpolator()
        ret.addAnimation(anim)
        anim = ScaleAnimation(1f, 0f, 1f, 0f, Animation.RELATIVE_TO_SELF, 0.5f, Animation.RELATIVE_TO_SELF, 0.5f)
        anim.duration = MEDIUM
        anim.interpolator = DecelerateInterpolator()
        ret.addAnimation(anim)
        return ret
    }

    @JvmStatic
    fun createPanInAnim(degree: Float): Animation {
        val ret = AnimationSet(false)
        var anim: Animation = AlphaAnimation(0f, 1f)
        anim.duration = MEDIUM
        anim.interpolator = LinearInterpolator()
        ret.addAnimation(anim)
        val pivotX = (1 - Math.cos(degree.toDouble())).toFloat() / 2
        val pivotY = (1 + Math.sin(degree.toDouble())).toFloat() / 2
        anim = ScaleAnimation(0.8f, 1f, 0.8f, 1f, Animation.RELATIVE_TO_SELF, pivotX, Animation.RELATIVE_TO_SELF, pivotY)
        anim.duration = MEDIUM
        anim.interpolator = DecelerateInterpolator()
        ret.addAnimation(anim)
        return ret
    }

    @JvmStatic
    fun createPanOutAnim(degree: Float): Animation {
        val ret = AnimationSet(false)
        var anim: Animation = AlphaAnimation(1f, 0f)
        anim.duration = MEDIUM
        anim.interpolator = DecelerateInterpolator()
        ret.addAnimation(anim)
        val pivotX = (1 + Math.cos(degree.toDouble())).toFloat() / 2
        val pivotY = (1 - Math.sin(degree.toDouble())).toFloat() / 2
        anim = ScaleAnimation(1f, 0.8f, 1f, 0.8f, Animation.RELATIVE_TO_SELF, pivotX, Animation.RELATIVE_TO_SELF, pivotY)
        anim.duration = MEDIUM
        anim.interpolator = DecelerateInterpolator()
        ret.addAnimation(anim)
        return ret
    }
}
